### PR TITLE
Fixed #25180 -- ArrayFields should not have varchar_patterns_ops or text_patterns_ops indexes

### PR DESCRIPTION
**Pull request by caioariede**
https://code.djangoproject.com/ticket/25180

### CODE REVIEW

**Comment by charettes** [comment]
This looks like an appropriate patch but it would require tests to make sure the indexes are not created.

**Comment by caioariede** [comment]
You're right. Just added tests.

**Comment by fabianbuechler** [comment]
What about array fields with defined length? Those look like `db_type='varchar[10]`, don't they?
`not db_type.endswith('[]')` won't catch that, I suppose.

File: django/db/backends/postgresql/schema.py [hunk 1]
```diff
@@ -29,10 +29,10 @@ def _model_indexes_sql(self, model):
                 # a second index that specifies their operator class, which is
                 # needed when performing correct LIKE queries outside the
                 # C locale. See #12234.
-                if db_type.startswith('varchar'):
+                if db_type.startswith('varchar') and not db_type.endswith('[]'):
```
Comment (coldmind) [inline django/db/backends/postgresql/schema.py]: Comment above also needs to be updated to not confuse reader about new adding, which is not related to ticket in comment

**Comment by caioariede** [comment]
Good catch! Patch updated.

File: tests/postgres_tests/test_array.py [hunk 2]
```diff
@@ -307,6 +307,28 @@ def test_adding_field_with_default(self):
...
             self.assertNotIn(table_name, connection.introspection.table_names(cursor))
 
+    @override_settings(MIGRATION_MODULES={
+        "postgres_tests": "postgres_tests.array_index_migrations",
+    })
+    def test_adding_arrayfield_with_index(self):
+        table_name = 'postgres_tests_chartextarrayindexmodel'
+        call_command('migrate', 'postgres_tests', verbosity=0)
+        with connection.cursor() as cursor:
+            conslike = [
```
Comment (timgraham) [inline tests/postgres_tests/test_array.py]: conslike -> like_constraint_field_names?
also, assertions could be moved out of the "with" statements.

File: tests/postgres_tests/test_array.py [hunk 3]
```diff
@@ -307,6 +307,28 @@ def test_adding_field_with_default(self):
         with connection.cursor() as cursor:
             self.assertNotIn(table_name, connection.introspection.table_names(cursor))
 
+    @override_settings(MIGRATION_MODULES={
+        "postgres_tests": "postgres_tests.array_index_migrations",
+    })
+    def test_adding_arrayfield_with_index(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_array.py]: Please add a docstring:
```
"""
ArrayField shouldn't have varchar_patterns_ops or text_patterns_ops indexes.
"""
```

File: django/db/backends/postgresql/schema.py [hunk 4]
```diff
@@ -29,6 +29,11 @@ def _model_indexes_sql(self, model):
                 # a second index that specifies their operator class, which is
                 # needed when performing correct LIKE queries outside the
                 # C locale. See #12234.
+                #
+                # The same does not apply to array fields such as varchar[size]
+                # and text[size], so we need to skip them
```
Comment (timgraham) [inline django/db/backends/postgresql/schema.py]: add period

**Comment by timgraham** [comment]
merged in dad8434d6ff5da10959672726dc9b397296d380b, thanks!

**Comment by fabianbuechler** [comment]
Thank you @caioariede for the patch and @timgraham for merging it!
Once this is released I can remove my crappy workaround :+1:

**Pull request closed without merging**